tools/aws/rds_backup_tool/image/rdsbackup.sh

```diff
@@ -0,0 +1,146 @@
+#!/bin/bash -e
+
+# env vars:
+# DBTYPE: accepts either of the following values: MYSQL or PGSQL
+# DBNAME: name of database to use in connection string
+# DBUSER: MySQL OR Postgres username
+# DBPASSWORD: MySQL OR Postgres password
+# DBHOST: MySQL OR Postgres host without port#
+# DBPORT: MySQL OR Postgres port #
+# BACKUP_CMD_PARAMS: additional parameters to pass to backup command for psql/mysql
+# BACKUP_BUCKET: where to write backup file, includes s3:// prefix
+# BACKUP_DIR: directory in the container mapped to EBS volume
+# BACKUP_PASSWORD: symmetric encryption password
+# AWS_ACCESS_KEY_ID
+# AWS_SECRET_ACCESS_KEY
+
+check_requirements() {
+    if [[ -z $DBHOST || -z $DBNAME || -z $DBPASSWORD \
+          || -z $DBPORT || -z $DBTYPE || -z $DBUSER || -z $BACKUP_DIR \
+          || -z $BACKUP_PASSWORD ]]; then
+        echo "ERROR: Not all environment variables are set."
+        echo "Minimum required vars:"
+        echo "  DBHOST"
+        echo "  DBNAME"
+        echo "  DBPASSWORD"
+        echo "  DBPORT"
+        echo "  DBTYPE"
+        echo "  DBUSER"
+        echo "  BACKUP_DIR"
+        echo "  BACKUP_PASSWORD"
+        exit 1
+    fi
+}
+
+fix_creds() {
+    # use a ~/.aws/credentials file INSTEAD of k8s set environment vars.
+    # Unset env vars after the file is written.
+    # Creds containing a slash seem to break the AWS CLI when run inside k8s,
+    # but running in docker is fine.
+    #
+    # example error:
+    # fatal error: Invalid header value b'AWS AKIAXXXX\n:ri81t84+26rX0qzRAASDDDXXXy2B70='
+    mkdir "${HOME}"/.aws
+    cat << EOF > "${HOME}"/.aws/credentials
+[default]
+aws_access_key_id=${AWS_ACCESS_KEY_ID}
+aws_secret_access_key=${AWS_SECRET_ACCESS_KEY}
+EOF
+    unset AWS_ACCESS_KEY_ID
+    unset AWS_SECRET_ACCESS_KEY
+}
+
+push_to_s3() {
+    REMOTE_DIR="${BACKUP_BUCKET}/${DBNAME}/"
+    echo "Pushing ${ENCRYPTED_BACKUP_FILENAME} to ${REMOTE_DIR}"
+    aws s3 cp "${ENCRYPTED_BACKUP_FILENAME}" "${REMOTE_DIR}"
+    echo "Finished pushing to S3"
+}
+
+encrypt_backup() {
+    echo "Encrypting ${BACKUP_FILENAME}"
+    echo -n "${BACKUP_PASSWORD}" | \
+        gpg --batch \
+            --passphrase-fd 0 \
+            --symmetric \
+            --cipher-algo aes256 \
+            ${BACKUP_FILENAME}
+    echo "Finished encrypting ${BACKUP_FILENAME}"
+    echo "Removing original: ${BACKUP_FILENAME}"
+    rm ${BACKUP_FILENAME}
```

Comment:
Why are you using an intermediate file instead of putting the encryption process in the pipeline between the db backup process the initial output to disk?

Reply:
because I have to pipe the password in on stdin to prevent the password prompt from appearing.

Reply:
I'm going to experiment with a pipeline that doesn't leave any unencrypted data on disk.